cpp/src/strings/split/split.cu

```diff
@@ -0,0 +1,526 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/split/split.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/utilities/error.hpp>
+#include <strings/utilities.hpp>
+
+#include <vector>
+#include <thrust/transform.h>
+#include <thrust/extrema.h>
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+
+using string_index_pair = thrust::pair<const char*,size_type>;
+
+namespace
+{
+
+/**
+ * @brief Common token counter for all split methods in this file.
+ */
+struct token_counter_fn
+{
+    column_device_view const d_strings;
+    string_view const d_delimiter;
+    size_type tokens;
+
+    // returns the number of possible tokens in each string
+    __device__ size_type operator()(size_type idx) const
+    {
+        if( d_strings.is_null(idx) )
+            return 0;
+        string_view d_str = d_strings.element<string_view>(idx);
+        if( d_str.empty() )
+            return 1;
+        size_type delim_count = 0;
+        auto delim_length = d_delimiter.length();
+        auto pos = d_str.find(d_delimiter);
+        while(pos >= 0)
+        {
+            ++delim_count;
+            pos = d_str.find(d_delimiter, pos + delim_length);
+        }
```

Comment:
We don't need to do this in this PR, but this is really an algorithm that should be pulled out into a "algorithm" header. I would call it `count_subsequence` - it is very similar to `count` however it counts sequences of elements (in this case a delimiter) instead of just a single element. It could be based on a generic version of the `string` / `string_view` `find` methods, which I would call `find_subsequence`. Sample (non-production versions i.e. we would actually want `difference_type` returned for the count algorithm) are below: 

```cpp 
template <typename I, typename I2>
auto find_subsequence(I f, I l, I2 f2, I2 l2) {
    auto sz = std::distance(f2, l2);
    while(sz <= std::distance(f, l)) {
        if (std::equal(f, f + sz, f2, l2)) 
            return f;
        ++f;
    }
    return l;
}

template <typename I, typename I2>
auto count_subsequence(I f, I l, I2 f2, I2 l2) {
    auto sz    = std::distance(f2, l2);
    auto count = 0;
    auto pos   = find_subsequence(f, l, f2, l2);
    while(pos != l) {
        ++count;
        pos = find_subsequence(pos + sz, l, f2, l2);
    }
    return count;
}
```
and then we could replace the highlight code with:
```cpp
size_type delim_count = count_subsequence(
        std::cbegin(d_str), 
        std::cend(d_str),
        std::cbegin(d_delimiter),
        std::cend(d_delimiter));
```
https://godbolt.org/z/mGF5TZ

Reply:
Thanks for this Conor. This code is actually from PR #3364 which this PR is dependent on. Regardless, this looks like an interesting approach. The problem is that `std::distance` may be an issue for `cudf::string_view` because `bytes != characters` and so simple pointer arithmetic would not work (i.e. random-access-iterator). The difference would need to be in characters and calculating this from the internal bytes could be expensive for longer strings.
We could profile against the current code and look at as an improvement in a future a PR perhaps if that is ok.

Reply:
Why does this need to be factored out into a *header*? Is it used a lot, or even anywhere outside this feature? Does it need to be supported for many types?